File: javascript/javascript/JavaScript/JavaScriptLexerBase.js
import antlr4 from 'antlr4';
import JavaScriptLexer from './JavaScriptLexer.js';

export default class JavaScriptLexerBase extends antlr4.Lexer {

    constructor(input) {
        super(input);
        this.scopeStrictModes = new Array();
        this.lastToken = null;
        this.useStrictDefault = false;
        this.useStrictCurrent = false;
        this.templateDepth = 0;
    }

    getStrictDefault() {
        return this.useStrictDefault;
    }

    setUseStrictDefault(value) {
        this.useStrictDefault = value;
        this.useStrictCurrent = value;
    }

    IsStrictMode() {
        return this.useStrictCurrent;
    }

    IsInTemplateString() {
        return this.templateDepth > 0;
    }

    getCurrentToken() {
        return this.nextToken();
    }

    nextToken() {
        var next = super.nextToken();

        if (next.channel === antlr4.Token.DEFAULT_CHANNEL) {
            this.lastToken = next;
        }
        return next;
    }

    ProcessOpenBrace() {
        this.useStrictCurrent =
            this.scopeStrictModes.length > 0 && this.scopeStrictModes[this.scopeStrictModes.length - 1]
                ? true
                : this.useStrictDefault;
        this.scopeStrictModes.push(this.useStrictCurrent);
    }

    ProcessCloseBrace() {
        this.useStrictCurrent =
            this.scopeStrictModes.length > 0
                ? this.scopeStrictModes.pop()
                : this.useStrictDefault;
    }

    ProcessStringLiteral() {
        if (this.lastToken === null ||
                this.lastToken.type === JavaScriptLexer.OpenBrace) {
            if (super.text === '"use strict"' || super.text === "'use strict'") {
                if (this.scopeStrictModes.length > 0) {
                    this.scopeStrictModes.pop();
                }
                this.useStrictCurrent = true;
                this.scopeStrictModes.push(this.useStrictCurrent);
            }
        }
    }

    IncreaseTemplateDepth() {
        this.templateDepth++;
    }

    DecreaseTemplateDepth() {
        this.templateDepth--;
    }

    IsRegexPossible() {
        if (this.lastToken === null) {
            return true;
        }

        switch (this.lastToken.type) {
            case JavaScriptLexer.Identifier:
            case JavaScriptLexer.NullLiteral:
            case JavaScriptLexer.BooleanLiteral:
            case JavaScriptLexer.This:
            case JavaScriptLexer.CloseBracket:
            case JavaScriptLexer.CloseParen:
            case JavaScriptLexer.OctalIntegerLiteral:
            case JavaScriptLexer.DecimalLiteral:
            case JavaScriptLexer.HexIntegerLiteral:
            case JavaScriptLexer.StringLiteral:
            case JavaScriptLexer.PlusPlus:
            case JavaScriptLexer.MinusMinus:
                return false;
            default:
                return true;
        }
    }

    IsStartOfFile() {
        return this.lastToken === null;
    }

    reset() {
        this.scopeStrictModes = new Array();
        this.lastToken = null;
        this.useStrictDefault = false;
        this.useStrictCurrent = false;
        this.templateDepth = 0;
        super.reset();
    }
}

File: javascript/javascript/Cpp/JavaScriptLexerBase.cpp
#include "JavaScriptLexer.h"

using namespace antlr4;

bool JavaScriptLexerBase::getStrictDefault()
{
    return useStrictDefault;
}

bool JavaScriptLexerBase::IsStartOfFile(){
    // No token has been produced yet: at the start of the input,
    // no division is possible, so a regex literal _is_ possible.
    return !lastToken;
}

void JavaScriptLexerBase::setUseStrictDefault(bool value)
{
    useStrictDefault = value;
    useStrictCurrent = value;
}

bool JavaScriptLexerBase::IsStrictMode()
{
    return useStrictCurrent;
}

bool JavaScriptLexerBase::IsInTemplateString()
{
	return _templateDepth > 0;
}

std::unique_ptr<antlr4::Token> JavaScriptLexerBase::nextToken() {
    auto next = Lexer::nextToken();

    if (next->getChannel() == Token::DEFAULT_CHANNEL) {
        // Keep track of the last token on the default channel.
        lastToken = true;
        lastTokenType = next->getType();
    }

    return next;
}

void JavaScriptLexerBase::ProcessOpenBrace()
{
    useStrictCurrent = scopeStrictModes.size() > 0 && scopeStrictModes.top() ? true : useStrictDefault;
    scopeStrictModes.push(useStrictCurrent);
}

void JavaScriptLexerBase::ProcessCloseBrace()
{
    if (scopeStrictModes.size() > 0) {
        useStrictCurrent = scopeStrictModes.top();
        scopeStrictModes.pop();
    } else {
        useStrictCurrent = useStrictDefault;
    }
}

void JavaScriptLexerBase::ProcessStringLiteral()
{
    if (!lastToken || lastTokenType == JavaScriptLexer::OpenBrace)
    {
        std::string text = getText();
        if (text == "\"use strict\"" || text == "'use strict'")
        {
            if (scopeStrictModes.size() > 0)
                scopeStrictModes.pop();
            useStrictCurrent = true;
            scopeStrictModes.push(useStrictCurrent);
        }
    }
}

void JavaScriptLexerBase::IncreaseTemplateDepth()
{
	_templateDepth++;
}

void JavaScriptLexerBase::DecreaseTemplateDepth()
{
	_templateDepth--;
}

bool JavaScriptLexerBase::IsRegexPossible()
{
    if (!lastToken) {
        // No token has been produced yet: at the start of the input,
        // no division is possible, so a regex literal _is_ possible.
        return true;
    }
    
    switch (lastTokenType) {
        case JavaScriptLexer::Identifier:
        case JavaScriptLexer::NullLiteral:
        case JavaScriptLexer::BooleanLiteral:
        case JavaScriptLexer::This:
        case JavaScriptLexer::CloseBracket:
        case JavaScriptLexer::CloseParen:
        case JavaScriptLexer::OctalIntegerLiteral:
        case JavaScriptLexer::DecimalLiteral:
        case JavaScriptLexer::HexIntegerLiteral:
        case JavaScriptLexer::StringLiteral:
        case JavaScriptLexer::PlusPlus:
        case JavaScriptLexer::MinusMinus:
            // After any of the tokens above, no regex literal can follow.
            return false;
        default:
            // In all other cases, a regex literal _is_ possible.
            return true;
    }
}

void JavaScriptLexerBase::reset()
{
    while(!scopeStrictModes.empty()) scopeStrictModes.pop();
    lastToken = false;
    lastTokenType = 0;
    useStrictDefault = false;
    useStrictCurrent = false;
    _templateDepth = 0;
    Lexer::reset();
}